File: Add Two Numbers/add.js

function ListNode(val) {
    this.val = val
    this.next = null
}
/**
 * 
 * @param {Array} sum 
 */
var NumberToNode = function NumberToNode(sum) {
    sum = sum.reverse();
    var prev = null;
    var next = null;
    if (sum && sum.length) {
        var n = Number(sum[0]);
        prev = new ListNode(n);
        for (var index = 1; index < sum.length; index++) {
            n = Number(sum[index])
            next = new ListNode(n);
            next.next = prev;
            prev = next;
        }

    }
    return prev

}

var l = [1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]
var s = [5, 6, 4]
var list1 = NumberToNode(l)
var list2 = NumberToNode(s)

var addTwoNumbers = function (l1, l2) {
    var dummyHead = new ListNode(0);
    var p = l1, q = l2; curr = dummyHead;
    var carry = 0; //为了进位
    var sum = 0;
    while (p || q) {
        var x = p ? p.val : 0;
        var y = q ? q.val : 0;
        sum = carry + x + y;
        carry = Math.floor(sum / 10);
        curr.next = new ListNode(sum % 10);
        curr = curr.next;
        if (p) p = p.next;
        if (q) q = q.next;
    }
    if (carry > 0) {
        curr.next = new ListNode(carry);
    }
    return dummyHead.next;
}


var node = addTwoNumbers(list1, list2)

logNode(node)
function logNode(node) {
    while (node) {
        console.log("------", node.val)
        node = node.next
    }
}

File: Add Two Numbers/add.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */

struct ListNode
{
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL){};
};

class Solution
{
  public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode *temp = new ListNode(0);
        ListNode *p = l1;
        ListNode *q = l2;
        ListNode *curr = temp;
        int carry = 0;
        int sum = 0;
        while (p != NULL || q != NULL)
        {
            int x = 0;
            int y = 0;
            if (p != NULL)
            {
                x = p->val;
                p = p->next;
            }
            if (q != NULL)
            {
                y = q->val;
                q = q->next;
            }

            sum = carry + x + y;
            carry = sum / 10; //是为了大于10 进1
            curr->next = new ListNode(sum % 10);
            curr = curr->next;
        }

        if (carry > 0)
        {
            curr->next = new ListNode(carry);
        }
        return temp->next;
    }
};